File: go/2015/day05b/day05b.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <input file>\n", progname)
	os.Exit(1)
}

func prop1(str string) bool {
	for outerStart := 0; outerStart < len(str)-3; outerStart++ {
		firstPair := str[outerStart : outerStart+2]
		for innerStart := outerStart + 2; innerStart < len(str)-1; innerStart++ {
			secondPair := str[innerStart : innerStart+2]
			if firstPair == secondPair {
				return true
			}
		}
	}
	return false
}

func prop2(str string) bool {
	for i := 0; i < len(str)-2; i++ {
		if str[i] == str[i+2] {
			return true
		}
	}
	return false
}

func process(filename string) uint32 {
	var result uint32 = 0

	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if prop1(line) && prop2(line) {
			result++
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	return result
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: fortran/2015/day05b/day05b.f90
PROGRAM day05b

  USE, INTRINSIC :: ISO_FORTRAN_ENV
  IMPLICIT NONE

  CHARACTER(LEN=:), ALLOCATABLE :: filename
  INTEGER :: filename_len, result

  IF ( COMMAND_ARGUMENT_COUNT() < 1 ) THEN
    CALL usage()
  END IF
  CALL GET_COMMAND_ARGUMENT(1, LENGTH=filename_len)
  ALLOCATE(CHARACTER(LEN=filename_len) :: filename)
  CALL GET_COMMAND_ARGUMENT(1, VALUE=filename)

  result = process(filename)
  WRITE (*,'(A,I0)') 'result = ', result
 
CONTAINS

  SUBROUTINE usage()
    CHARACTER(LEN=:), ALLOCATABLE :: progname
    INTEGER :: progname_len

    CALL GET_COMMAND_ARGUMENT(0, LENGTH=progname_len)
    ALLOCATE(CHARACTER(LEN=progname_len) :: progname)
    CALL GET_COMMAND_ARGUMENT(0, VALUE=progname)
    WRITE (error_unit, *) 'usage: ' // TRIM(progname) // ' <input file>' 
    CALL EXIT(1)
  END SUBROUTINE usage

  FUNCTION prop1(str) RESULT(res)
    CHARACTER(LEN=32), INTENT(IN) :: str
    CHARACTER(LEN=2) :: first_pair
    LOGICAL :: res
    INTEGER :: i

    DO i = 1, LEN_TRIM(str) - 3
      first_pair = str(i:i+1)
      IF ( INDEX(str(i+2:), first_pair) /= 0 ) THEN
        res = .TRUE.
        RETURN
      END IF
    END DO
    res = .FALSE.
  END FUNCTION prop1

  FUNCTION prop2(str) RESULT(res)
    CHARACTER(LEN=32), INTENT(IN) :: str
    LOGICAL :: res
    INTEGER :: i

    DO i = 3, LEN_TRIM(str)
      IF ( str(i:i) == str(i-2:i-2) ) THEN
        res = .TRUE.
        RETURN
      END IF
    END DO
    res = .FALSE.
  END FUNCTION prop2

  FUNCTION process(filename) RESULT(count)
    CHARACTER(LEN=*), INTENT(IN) :: filename
    INTEGER :: count
    INTEGER :: file_unit, ios
    CHARACTER(LEN=32) :: line

    count = 0

    OPEN(NEWUNIT=file_unit, &
         FILE=TRIM(filename), &
         STATUS='OLD', &
         ACTION='READ', &
         IOSTAT=ios)
    IF ( ios /= 0 ) THEN
      WRITE(error_unit, *) "Error opening file: " // filename
      CALL EXIT(1)
    END IF

    DO
      READ(file_unit, '(A)', iostat=ios) line
      IF ( ios /= 0 ) EXIT
      IF ( prop1(line) .AND. prop2(line) ) THEN
        count = count + 1
      END IF
    END DO

    CLOSE(file_unit)
  END FUNCTION process

END PROGRAM day05b
